File: backend-laravel/app/Models/MateriaTaller.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Database\Eloquent\SoftDeletes;

/**
 * Materia/taller de una especialidad (tabla `materias_taller`).
 * `regimen_cursada` es por materia, no global — la narrativa exige que
 * cada taller pueda tener su propia duración (anual/trimestral/
 * semestral/personalizado), a diferencia del curso principal que no
 * tiene este dato.
 *
 * `especialidad_id` es nullable: las escuelas técnico-profesionales
 * recién asignan la orientación a partir de 3°/4° año, así que las
 * materias de ciclo básico (1°/2° año) se cargan sin especialidad. Ver
 * la migración `make_especialidad_nullable_en_materias_taller`.
 */
class MateriaTaller extends Model
{
    use SoftDeletes;

    protected $table = 'materias_taller';

    protected $primaryKey = 'id_materia_taller';

    protected $fillable = [
        'especialidad_id',
        'nombre',
        'regimen_cursada',
    ];

    protected function casts(): array
    {
        return [
            'deleted_at' => 'datetime',
        ];
    }

    public function especialidad(): BelongsTo
    {
        return $this->belongsTo(Especialidad::class, 'especialidad_id', 'id_especialidad');
    }

    public function gruposTaller(): HasMany
    {
        return $this->hasMany(GrupoTaller::class, 'materia_taller_id', 'id_materia_taller');
    }
}

File: app_flutter/lib/models/materia_taller.dart
/// Espejo de lo que devuelve `MateriasTallerController` (backend-laravel)
/// — una materia/taller del catálogo de una especialidad (ej. "Dibujo
/// Técnico" en Electromecánica). Dato permanente, no depende de ningún
/// ciclo lectivo — los grupos de taller (`GrupoTaller`) sí son por
/// ciclo y se arman aparte, sobre estas materias ya existentes. Ver
/// `App\Models\MateriaTaller` en el backend.
///
/// `especialidadId`/`especialidadNombre` son nullable: una materia de
/// ciclo básico (1°/2° año) puede no tener especialidad todavía, ya que
/// en las escuelas técnico-profesionales la orientación recién se define
/// a partir de 3°/4° año.
class MateriaTaller {
  const MateriaTaller({
    required this.id,
    required this.especialidadId,
    required this.especialidadNombre,
    required this.nombre,
    required this.regimenCursada,
  });

  factory MateriaTaller.fromJson(Map<String, dynamic> json) {
    return MateriaTaller(
      id: json['id_materia_taller'] as int,
      especialidadId: json['especialidad_id'] as int?,
      especialidadNombre: json['especialidad_nombre'] as String?,
      nombre: json['nombre'] as String,
      regimenCursada: json['regimen_cursada'] as String,
    );
  }

  final int id;
  final int? especialidadId;
  final String? especialidadNombre;
  final String nombre;

  /// 'anual' | 'trimestral' | 'semestral' | 'personalizado'.
  final String regimenCursada;
}
